### ISSUE TITLE
Convert TestKitBase.testActor to a def + private lazy val

### ISSUE THREAD

**Issue by spangaer** [akka/akka#23495]
Currently TestKitBase.testActor is created as a val of TestKitBase at construction time.
This could be converted to a def + private lazy val.
The private lazy val would allow to keep the current default behavior available.
The def would allow derivatives of TestKitBase to override the Actor creation behavior.
The message queue will probably have to be made protected as well.

This opens the door to create a TestActor as a child actor to "a real actor" and let that TestActor create the real child. Allowing to intercept parent - child communication, without breaking up that relationship, but still have all the TestKit utilities available.
(You still need a forwarding autopilot to complete the story, but that's possible with how it works today).

**Comment by johanandren** [comment]
As the `TestActor` is only for synchronous testing (clearly noted with a warning box in the docs here: http://doc.akka.io/docs/akka/current/scala/testing.html#synchronous-testing-testactorref ) I don't think it really makes sense to allow it to take part in a hierarchy of real actors. For such use cases the recommended solution is traditional behavioural testing, also noted on that doc page.

**Comment by spangaer** [comment]
Isn't that warning talking about TestActorRef, instead of TestActor?
Seems like TestActor has no synchronous requirements, to extend on that, the explicit usage off a concurrent queue suggests the opposite.
I've made a hack so far, I'm still on my way to make it work, as soon as I get there I'll post some snippets.

**Comment by johanandren** [comment]
You are entirely right, sorry, my brain added a `Ref` that wasn't there.

I'm wondering if this elaborate behaviour is really something to encourage on top of a concept that is otherwise relatively simple? Isn't it good that tests testing parent child messaging explicitly has to express that? 

Note that all the facilities of the test actor are available through `TestProbe` instances, which you can create any number of.

**Comment by spangaer** [comment]
So a little bit of context on what I'm trying to do.
I have a set of 3-4 Akka Persistence `Persistent Actors`, some of the `PersistentFSM`.
Some of them have a Parent-Child relationship others don't.
With this constellation of Actors I'm trying to bring a minimal form of "transactonality/journalling" of modifications that span multiple persistent nodes. The goal is, either the constellation succeeds in going in to a transactional state and then always completes the operation successfully or fails to do so and reverts to what it was.

I appreciate that that might go a bit against the spirit of the whole even sourcing and yes for many things we will probably successfully avoid the need for this altogether, but not for everything.

Now I have this constellation that sort of works as expected, how do I really know that it does. You could indeed isolate all of the actors and give them expected inputs and check outputs. But the interaction pattern is much complexer then simple parent child, I would be required to stub my brains out and I need to run the `Persistent Actors` in context. That while for a unit testing environment I can just use an in memory journal plugin, with the real constellation.

So, I've introduced an Instrumentation Akka Extension ("abusing" it as app singleton) that is called for actor construction of interesting actors.
In standard deployment it just calls through to standard `actorOf`. In testing conditions I can configure it to replace certain construction or lookup by test probes.
So I get one `probe` pretending as a child but forwarding to the real child and vica versa.

Again using the Instrumentation singleton I can fetch the probes and observe the communication. But soon I want to use the mechanism for message stalling and fault injection (PoisonPill).

Thus I've added following clone to achieve my goal.

```scala
trait ETestKitBase {

  ...

  def testActor = _testActor

  private lazy val _testActor: ActorRef = {
    val impl = system.asInstanceOf[ExtendedActorSystem]
    val ref = impl.systemActorOf(
      TestActor.props(queue)
        .withDispatcher(CallingThreadDispatcher.Id),
      "%s-%d".format(testActorName, TestKit.testActorId.incrementAndGet))
    awaitCond(ref match {
      case r: RepointableRef ⇒ r.isStarted
      case _ ⇒ true
    }, 3.seconds.dilated, 10.millis)
    ref
  }
  
  ...
}
class ETestKit(_system: ActorSystem) extends { implicit val system = _system } with ETestKitBase


class EChildTestProbe(context: ActorContext, nameO: Option[String] = None) extends ETestKit(context.system) {

  override val testActor = {
    val ref = context.actorOf(TestActor.props(queue), testActorName)
    awaitCond(ref match {
      case r: RepointableRef ⇒ r.isStarted
      case _ ⇒ true
    }, 3.seconds.dilated, 10.millis)
    ref
  }

  override def testActorName = nameO.getOrElse {
    "%s-%d".format("childTestProbe", TestKit.testActorId.incrementAndGet)
  }
  
  ...

}
```

Which is wired together with a forwarding autopilot:
```scala
private class OneWayForwardingAutoPilot(dest: ActorRef, senderReplace: Option[(ActorRef, { def ref: ActorRef })]) extends TestActor.AutoPilot {

  override def run(sender, msg) = {
    //     
    senderReplace match {
      case Some((toReplace, replacement)) if (toReplace == sender) =>
        dest.tell(msg, replacement.ref)
      case _ =>
        dest.tell(msg, sender)
    }

    keepRunning
  }

}
```

```scala
      val childProbe = new EChildTestProbe(context, name)
      val parentProbe = TestProbe(childProbe.testActorName)(context.system)
      childProbe.watch(parentProbe.ref)
      parentProbe.watch(childProbe.ref)

      val realParent = context.self
      val realChild = parentProbe.childActorOf(props, childProbe.testActorName)

      childProbe.setAutoPilot(new OneWayForwardingAutoPilot(realChild, Some((realParent, parentProbe))))
      parentProbe.setAutoPilot(new OneWayForwardingAutoPilot(realParent, Some((realChild, childProbe))))
```

**Comment by spangaer** [comment]
I guess a potato was in order, to signal the end of the long post and not smack down the conversation?

**Comment by spangaer** [comment]
Well, after 2 iterations (with some holiday interleaved) of trying it that way, I ended up highly unsatisfied with how it turned out to be working. So put on some modern engineering boots and tried again with some 'inversion of control'/'aspect oriented' style (if you will)

- I've kept the `Instumentation` extension as `ActorSystem` singleton and with configurable behaviour: production has a no-op implemention, but it can be swapped to some more advanced stuff specifically for testing.
- A trait extends `Actor` and overrides `aroundReceive()` and passes through the extension with `super.aroundReceive(_, _)` as closure.
- Again in production no-op, but with an alternative testing version of the extension for certain actors it can activate a series of message triggered fault injections, simulating: Lost messages, Unexpected Termination,...

So with that I am indeed indeed landing at: "external observation (out of system)" but with "internal fault simulation".
I ended here, because:

- Trying the the same with the test probe approach became unmanageable quickly. (e.g. forwarding sometimes injecting error at other times)
- This way I don't have to make any 'real' code aware of this whole instrumentation story, it's handled at a much higher level.
